bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -187,155 +202,139 @@ public RefreshingOAuth2CredentialsInterceptor(ExecutorService scheduler,
-      if (!isRefreshing) {
-        doRefresh();
-      } else {
-        while (isRefreshing && getCacheState(this.headerCache.get()) != CacheState.Good) {
-          try {
-            lock.wait(250);
-          } catch (InterruptedException e) {
-            Thread.currentThread().interrupt();
-            throw new IOException(e);
-          }
-        }
-      }
+  HeaderCacheElement syncRefresh() {
+    try {
+      return asyncRefresh().get(250, TimeUnit.MILLISECONDS);
+    } catch (Exception e) {
+      return new HeaderCacheElement(
+          Status.UNAUTHENTICATED
+              .withCause(e)
+      );
     }
   }
 
   /**
    * Refreshes the OAuth2 token asynchronously. This method will only start an async refresh if
-   * there isn't a currently running asynchronous refresh or the credentials are for App Engine.
-   * App Engine has some credentials related state in a {@link ThreadLocal} which prevents it from
-   * running asynchronously. In the App Engine case, this method will defer to
-   * {@link #syncRefresh()}.
-   */
-  void asyncRefresh() throws IOException {
-    if (isAppEngine) {
-      syncRefresh();
-    } else if (canRefresh()) {
-      executor.execute(new Runnable() {
-        @Override
-        public void run() {
-          doRefresh();
-        }
-      });
-    }
-  }
-
-  private boolean canRefresh() {
-    synchronized (lock) {
-      return !isRefreshing && getCacheState(this.headerCache.get()) != CacheState.Good;
-    }
-  }
-
-  private static StatusRuntimeException asUnauthenticatedException(Exception e) {
-    return Status.UNAUTHENTICATED.withCause(e).asRuntimeException();
-  }
-
-  @VisibleForTesting
-  static CacheState getCacheState(HeaderCacheElement headerCache) {
-    return (headerCache == null) ? CacheState.Expired : headerCache.getCacheState();
-  }
-
-  private HeaderCacheElement getCachedHeader() throws IOException {
-    HeaderCacheElement headerCache = this.headerCache.get();
-    if (headerCache != null && headerCache.exception != null) {
-      throw headerCache.exception;
-    }
-    return headerCache;
-  }
-
-  /**
-   * Perform a credentials refresh.
+   * there isn't a currently running asynchronous refresh.
    */
-  @VisibleForTesting
-  boolean doRefresh() {
-    if (!canRefresh()) {
-      return false;
-    }
+  Future<HeaderCacheElement> asyncRefresh() {
     synchronized (lock) {
-      if (!canRefresh()) {
-        return false;
+      if (isRefreshing) {
+        return futureToken;
       }
       isRefreshing = true;
-    }
-    try {
-      HeaderCacheElement cacheElement = refreshCredentialsWithRetry();
-      synchronized (lock) {
-        updateToken(cacheElement);
-      }
-    } finally {
-      synchronized (lock) {
+
+      try {
+        this.futureToken = executor.submit(new Callable<HeaderCacheElement>() {
+          @Override
+          public HeaderCacheElement call() throws Exception {
+            HeaderCacheElement newToken = refreshCredentialsWithRetry();
```

Comment:
refreshCredentialsWithRetry() can throw a RuntimeException.  Please add a try-finally clause to set `isRefreshing = false`.  (see `getRetryState()`)